pandas/io/excel/_openpyxl.py

```diff
@@ -595,36 +590,74 @@ def _convert_cell(self, cell, convert_float: bool) -> Scalar:
         return cell.value
 
     def get_sheet_data(
-        self, sheet, convert_float: bool, file_rows_needed: int | None = None
+        self,
+        sheet,
+        convert_float: bool,
+        file_rows_needed: int | None = None,
+        offset: int | None = None,
     ) -> list[list[Scalar]]:
 
         if self.book.read_only:
             sheet.reset_dimensions()
 
-        data: list[list[Scalar]] = []
-        last_row_with_data = -1
-        for row_number, row in enumerate(sheet.rows):
-            converted_row = [self._convert_cell(cell, convert_float) for cell in row]
-            while converted_row and converted_row[-1] == "":
-                # trim trailing empty elements
-                converted_row.pop()
-            if converted_row:
-                last_row_with_data = row_number
-            data.append(converted_row)
-            if file_rows_needed is not None and len(data) >= file_rows_needed:
-                break
-
-        # Trim trailing empty rows
-        data = data[: last_row_with_data + 1]
-
-        if len(data) > 0:
-            # extend rows to max width
-            max_width = max(len(data_row) for data_row in data)
-            if min(len(data_row) for data_row in data) < max_width:
-                empty_cell: list[Scalar] = [""]
-                data = [
-                    data_row + (max_width - len(data_row)) * empty_cell
-                    for data_row in data
+        def _loop_rows(iterator):
+            data: list[list[Scalar]] = []
+            last_row_with_data = -1
+            for row_number, row in iterator:
+                converted_row = [
+                    self._convert_cell(cell, convert_float) for cell in row
                 ]
+                while converted_row and converted_row[-1] == "":
+                    # trim trailing empty elements
+                    converted_row.pop()
+                if converted_row:
+                    last_row_with_data = row_number
+                data.append(converted_row)
+                if file_rows_needed is not None and len(data) >= file_rows_needed:
+                    break
+
+            # Trim trailing empty rows
+            data = data[: last_row_with_data + 1]
+
+            if data:
+                # extend rows to max width
+                max_width = max(len(data_row) for data_row in data)
+                if min(len(data_row) for data_row in data) < max_width:
+                    empty_cell: list[Scalar] = [""]
+                    data = [
+                        data_row + (max_width - len(data_row)) * empty_cell
+                        for data_row in data
+                    ]
+
+            return data
+
+        data: list[list[Scalar]] = []
+        if file_rows_needed or offset:
```

Comment:
```suggestion
        data: list[list[Scalar]] = []
        file_rows_needed = 0 if file_rows_needed is None else file_rows_needed
       offset = 0 if offset is None else offset
        if file_rows_needed or offset:
```
This is suggested to simplify the logic bellow